src/librustc/middle/outlives.rs

```diff
@@ -0,0 +1,191 @@
+                // to the object type), we get the requirement that
+                // `'a: 'static` since `'a` appears in the closure
+                // type `C`.
+                //
+                // A smarter fix might "prune" unused `func_substs` --
+                // this would avoid breaking simple examples like
+                // this, but would still break others (which might
+                // indeed be invalid, depending on your POV). Pruning
+                // would be a subtle process, since we have to see
+                // what func/type parameters are used and unused,
+                // taking into consideration UFCS and so forth.
+
+                for &upvar_ty in &substs.upvar_tys {
+                    compute_components(infcx, upvar_ty, out);
+                }
+                subtys.skip_current_subtree();
+            }
+            ty::TyBareFn(..) | ty::TyTrait(..) => {
+                subtys.skip_current_subtree();
+                let temp = capture_components(infcx, ty);
+                out.push(Component::RFC1214(temp));
+            }
+            ty::TyParam(p) => {
+                out.push(Component::Param(p));
+                subtys.skip_current_subtree();
+            }
+            ty::TyProjection(ref data) => {
+                // For projections, we prefer to generate an
+                // obligation like `<P0 as Trait<P1...Pn>>::Foo: 'a`,
+                // because this gives the regionck more ways to prove
+                // that it holds. However, regionck is not (at least
+                // currently) prepared to deal with higher-ranked
+                // regions that may appear in the
+                // trait-ref. Therefore, if we see any higher-ranke
+                // regions, we simply fallback to the most restrictive
+                // rule, which requires that `Pi: 'a` for all `i`.
+
+                if !data.has_escaping_regions() {
+                    // best case: no escaping reions, so push the
+                    // projection and skip the subtree (thus
+                    // generating no constraints for Pi).
+                    out.push(Component::Projection(*data));
+                } else {
+                    // fallback case: continue walking through and
+                    // constrain Pi.
+                    let temp = capture_components(infcx, ty);
+                    out.push(Component::EscapingProjection(temp));
+                }
+                subtys.skip_current_subtree();
+            }
+            ty::TyInfer(_) => {
+                let ty = infcx.resolve_type_vars_if_possible(&ty);
+                if let ty::TyInfer(infer_ty) = ty.sty {
+                    out.push(Component::UnresolvedInferenceVariable(infer_ty));
+                } else {
+                    compute_components(infcx, ty, out);
+                }
+            }
+            _ => {
+                // for all other types, just constrain the regions and
```

Comment:
I prefer to use a match instead of`Ty::walk`. It isn't so much longer in these cases, and it makes it harder to miss interesting variants, especially when new ones are added :-).